autogen/token_count_utils.py

```diff
@@ -0,0 +1,155 @@
+import tiktoken
+from typing import List, Union, Dict, Tuple
+import logging
+import json
+
+
+logger = logging.getLogger(__name__)
+
+
+def get_max_token_limit(model="gpt-3.5-turbo-0613"):
+    max_token_limit = {
+        "gpt-3.5-turbo": 4096,
+        "gpt-3.5-turbo-0301": 4096,
+        "gpt-3.5-turbo-0613": 4096,
+        "gpt-3.5-turbo-16k": 16384,
+        "gpt-35-turbo": 4096,
+        "gpt-4": 8192,
+        "gpt-4-32k": 32768,
+        "gpt-4-32k-0314": 32768,  # deprecate in Sep
+        "gpt-4-0314": 8192,  # deprecate in Sep
+        "gpt-4-0613": 8192,
+        "gpt-4-32k-0613": 32768,
+    }
+    return max_token_limit[model]
+
+
+def percentile_used(input, model="gpt-3.5-turbo-0613"):
+    return count_token(input) / get_max_token_limit(model)
+
+
+def token_left(input, model="gpt-3.5-turbo-0613"):
+    return get_max_token_limit(model) - count_token(input, model=model)
+
+
+def count_token(input: Union[str, List, Dict], model: str = "gpt-3.5-turbo-0613"):
+    if isinstance(input, str):
+        return _num_token_from_text(input, model=model)
+    elif isinstance(input, list) or isinstance(input, dict):
+        return _num_token_from_messages(input, model=model)
+    else:
+        raise ValueError("input must be str, list or dict")
+
+
+def _num_token_from_text(text: str, model: str = "gpt-3.5-turbo-0613"):
+    try:
+        encoding = tiktoken.encoding_for_model(model)
+    except KeyError:
+        print("Warning: model not found. Using cl100k_base encoding.")
+        encoding = tiktoken.get_encoding("cl100k_base")
+    return len(encoding.encode(text))
+
+
+def _num_token_from_messages(messages: Union[List, Dict], model="gpt-3.5-turbo-0613", verbose=False):
```

Comment:
I'd suggest we break the token count functions into several sub functions:
func 1. count the tokens of a text string
func 2. count the tokens of a function call (func 1 + function overheads)
func 3. count the tokens of a message list (func 1 + func 2 + message overheads)

And, we need to support models other than openai models, providing a parameter for passing in cutomized token count function.

Here is what I used in retrieve_utils.py, it's not perfect, I think you can combine what you've here and those in retrieve_utils.py to make a solution. So we can all use your version in the future.

https://github.com/microsoft/autogen/blob/main/autogen/retrieve_utils.py#L36-L107

Reply:
@kevin666aa @sonichi

Reply:
If they need to pass in customized token count function, I suppose they can use it directly? If a different model is used, func 1-3 would all be different. Is there a case this would be necessary?

I will check again your solution and refine what I have now. Essentially these functions come from the same source so it is basically the same.